changefeedccl: db-level changefeeds: notice newly added tables

asg0451 opened the pull request:
Add a table watcher check to the change frontier
before saving progress. If we encounter a new
table, save progress up to that point and then
die.

Fixes: #148834
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152869)
<!-- Reviewable:end -->

asg0451 committed:
changefeedccl: db-level changefeeds: notice newly added tables

Add a table watcher check to the change frontier
before saving progress. If we encounter a new
table, save progress up to that point and then
die.

Fixes: #148834
Release note: None

asg0451 committed:
changefeedccl: make tableset watcher ignore databases and schemas

Previously the tableset watcher would incorrectly
report table drops for databases and schemas. This is now fixed.

Epic: none
Release note: None

asg0451 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1942,16 +2038,31 @@ func (cf *changeFrontier) maybePersistFrontier(ctx context.Context) error {
 	return nil
 }
 
+// TODO: this api is a bit sloppy; find a way to do it with its own and with a passed in txn better.
```
this

asg0451 committed:
changefeedccl: make config error terminal but not assertion-failed

If you specify a table in a database-level
changefeed that's in another database, it will
return a terminal error instead of an
assertion-failed error.

Epic: None
Release note: None

KeithCh commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1928,6 +2024,10 @@ func (cf *changeFrontier) checkpointJobProgress(
 	cf.localState.SetCheckpoint(spanLevelCheckpoint)
 
+	if len(tablesToAdd) > 0 {
+		return errors.Wrapf(errDatabaseTargetsChanged, "(progress saved) before checkpoint at %s: %v", frontier, tablesToAdd)
```
Can you elaborate on why we want to do this?

KeithCh commented on pkg/ccl/changefeedccl/tableset/watcher.go:
Maybe worth resolving this TODO in this PR.

asg0451 commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1873,6 +1924,44 @@ func (cf *changeFrontier) checkpointJobProgress(
+			// If there are changes, we'll only advance the frontier to the creation time of the first added table.
+			frontier = diffs.Frontier()
+			// Add new tables to the frontier and persist it, if there are any.
+			if len(tablesToAdd) > 0 {
```
what do you mean?

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 10.91m ±1% | 10.98m ±1% | ~ | p=0.174 n=15    |
| ⚪ **allocs/op** | 8.056k ±1% | 8.064k ±1% | ~ | p=0.395 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/1f40090/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/1f400905a76f09521a2914f64db75d27a8341591/bin/pkg_sql_tests benchdiff/1f40090/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/1f40090/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/ee20eae/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/ee20eaee6fe87d6a8597784b21bdf96469e763bc/bin/pkg_sql_tests benchdiff/ee20eae/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/ee20eae/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=ee20eae --new=1f40090 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>🔴 Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| 🔴 **sec/op** | 3.293m ±1% | 3.347m ±1% | +1.63% | p=0.003 n=15    |
| ⚪ **allocs/op** | 2.080k ±0% | 2.080k ±0% | ~ | p=0.453 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/1f40090/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/1f400905a76f09521a2914f64db75d27a8341591/bin/pkg_sql_tests benchdiff/1f40090/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/1f40090/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/ee20eae/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/ee20eaee6fe87d6a8597784b21bdf96469e763bc/bin/pkg_sql_tests benchdiff/ee20eae/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/ee20eae/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=ee20eae --new=1f40090 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.716m ±1% | 3.709m ±0% | ~ | p=0.486 n=15    |
| ⚪ **allocs/op** | 4.196k ±0% | 4.193k ±0% | ~ | p=0.050 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/1f40090/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/1f400905a76f09521a2914f64db75d27a8341591/bin/pkg_sql_tests benchdiff/1f40090/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/1f40090/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/ee20eae/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/ee20eaee6fe87d6a8597784b21bdf96469e763bc/bin/pkg_sql_tests benchdiff/ee20eae/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/ee20eae/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=ee20eae --new=1f40090 ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/1f400905a76f09521a2914f64db75d27a8341591/19109006954-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/ee20eaee6fe87d6a8597784b21bdf96469e763bc/19109006954-1/\* old/
```

</details>

_built with commit: [1f400905a76f09521a2914f64db75d27a8341591](https://github.com/cockroachdb/cockroach/commit/1f400905a76f09521a2914f64db75d27a8341591)_

<!-- microbench-ci -->
